<redacted> <redacted>

GitHub: <redacted>

https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

Commit ID: cfe46b94fae24ba37593d09f855ba8ba8b077956

In [2]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier


import warnings
warnings.filterwarnings("ignore")

# Part 1: Load the dataset

In [3]:
# Load the dataset (load remotely or locally)
df = pd.read_csv("iris.csv")

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

print(df.head(15))
print(df.describe())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

# Part 2: Split the dataset into train and test

In [4]:
# Take the dataset and split it into our features (X) and label (y)
X = df[["sepal_length", "sepal_width", "petal_length", "petal_width"]]
y = df["species"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=69)


# Part 3: Logistic Regression

In [5]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred = model.predict_proba(X_test[0:1])
print("prediction probabilities: ", [(model.classes_[i], y_pred[0][i]) for i in range(len(model.classes_))]) # pair each class with the corresponding probability
print("true label: ", y_test.iloc[0])

# iii. Report on the score for Logistic regression model, what does the score measure?
#score measures fraction of correct classifications
print("accuracy: ", model.score(X_test, y_test))

# iv. Extract the coefficents and intercepts for the boundary line(s)
coef = model.coef_
intercept = model.intercept_
print("coefficients: ", coef)
print("intercepts: ", intercept)

prediction probabilities:  [('setosa', 0.9842259507202842), ('versicolor', 0.015774017274534573), ('virginica', 3.200518136638976e-08)]
true label:  setosa
accuracy:  0.9333333333333333
coefficients:  [[-0.438494    0.89958177 -2.42436676 -1.04421628]
 [ 0.57126629 -0.47595953 -0.16182314 -0.9670602 ]
 [-0.13277229 -0.42362225  2.5861899   2.01127648]]
intercepts:  [  9.68191887   2.26858246 -11.95050133]


Part iii:

Score measures fraction of correct classifications.

Part iv:

There are 3 boundary lines; one for each classification type.
$x$ is a vector of [sepal length, sepal width, petal length, petal width]

$0 = 9.681 - 0.438x_1 + 0.899x_2 - 2.424x_3 - 1.0442x_4$

$0 = 2.268 + 0.571x_1 - 0.475x_2 - 0.161x_3 - 0.967x_4$

$0 = -11.950 - 0.132x_1 - 0.423x_2 + 2.586x_3 + 2.011x_4$


# Part 4: Support Vector Machine

In [6]:
# i. Use sklearn to train a Support Vector Classifier on the training set
model = SVC(probability=True)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred = model.predict_proba(X_test[0:1])
print("prediction probabilities: ", [(model.classes_[i], y_pred[0][i]) for i in range(len(model.classes_))])
print("true label: ", y_test.iloc[0])

# iii. Report on the score for the SVM, what does the score measure?
#score measures fraction of correct classifications
print("accuracy: ", model.score(X_test, y_test))

prediction probabilities:  [('setosa', 0.9762427893639372), ('versicolor', 0.014847841287994108), ('virginica', 0.008909369348068826)]
true label:  setosa
accuracy:  0.9333333333333333


Part iii:

Score measures fraction of correct classifications.

# Part 5: Neural Network

In [7]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
model = MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=1, activation="logistic", solver="adam")
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred = model.predict_proba(X_test[0:1])
print("prediction probabilities: ", [(model.classes_[i], y_pred[0][i]) for i in range(len(model.classes_))])
print("true label: ", y_test.iloc[0])

# iii. Report on the score for the Neural Network, what does the score measure?
print("accuracy: ", model.score(X_test, y_test))

# iv: Experiment with different options for the neural network, report on your best configuration
activations = ["relu", "logistic", "identity", "tanh"]
num_sizes = 4
hidden_sizes = np.logspace(0,num_sizes - 1,num_sizes,base=2, dtype=int)

#note: it seems fairly trivial to get 1.0 accuracy with most configurations with high hidden layer size
print("\n-------------")
for activation in activations:
    print(f'Models using {activation} activation:')
    for size in hidden_sizes:
        model = MLPClassifier(hidden_layer_sizes=(size,), max_iter=1000, random_state=1, activation=activation, solver="adam")
        model.fit(X_train, y_train)
        acc = model.score(X_test, y_test)
        print(f'NN with hidden layer of size {size}', "accuracy: ", acc)

prediction probabilities:  [('setosa', 0.9594206128936513), ('versicolor', 0.038006581549248186), ('virginica', 0.002572805557100503)]
true label:  setosa
accuracy:  1.0

-------------
Models using relu activation:
NN with hidden layer of size 1 accuracy:  0.2
NN with hidden layer of size 2 accuracy:  0.6
NN with hidden layer of size 4 accuracy:  0.9333333333333333
NN with hidden layer of size 8 accuracy:  1.0
Models using logistic activation:
NN with hidden layer of size 1 accuracy:  0.2
NN with hidden layer of size 2 accuracy:  0.6
NN with hidden layer of size 4 accuracy:  1.0
NN with hidden layer of size 8 accuracy:  1.0
Models using identity activation:
NN with hidden layer of size 1 accuracy:  0.4
NN with hidden layer of size 2 accuracy:  0.7333333333333333
NN with hidden layer of size 4 accuracy:  0.9333333333333333
NN with hidden layer of size 8 accuracy:  1.0
Models using tanh activation:
NN with hidden layer of size 1 accuracy:  0.2
NN with hidden layer of size 2 accuracy:  1.

Part iii:

Score measures fraction of correct classifications.

Part iv: 

I tested a few different hyperparameters: the activation function (relu, logistic, identity, and tanh), as well as the number of nodes in the hidden layer. All models only had 1 hidden layer.

Although all the models were able to reach a perfect 1.0 accuracy, only the tanh activation was able to reach it with a hidden size of 2. With the Occam's Razor principle (simpler models of equal performance are better), the best configuration is the tanh activation function with hidden size of 2.

# Part 6: K-Nearest Neighbors

In [8]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
k = 5  # Number of neighbors
model = KNeighborsClassifier(n_neighbors=k)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred = model.predict_proba(X_test[0:1])
print(y_pred)
print("prediction probabilities: ", [(model.classes_[i], y_pred[0][i]) for i in range(len(model.classes_))])


# iii. Report on the score for kNN, what does the score measure?
#score measures fraction of correct classifications
print("accuracy: ", model.score(X_test, y_test))

[[1. 0. 0.]]
prediction probabilities:  [('setosa', 1.0), ('versicolor', 0.0), ('virginica', 0.0)]
accuracy:  0.9333333333333333


Part iii:

Score measures fraction of correct classifications.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Logistic regression, SVM, and KNN Classifier performed with 0.93 accuracy, or mis-classifying 1/15 test examples. However, the neural network classifier got 1.0, which is a perfect score. In fact, the small NN of size 2 with tanh activation also got a perfect accuracy. This is because the NN has the highest capacity out of all the models, even with only 2 hidden nodes, so it is able to model a more complex boundary.

Surprisingly, the random seed in train_test_split had a strong and noticeable effect on a few of the accuracies, which indicates that the 0.1 test set is not enough to be representative of the data.

With random_state=1 in train_test_split, every single model had a perfect accuracy of 1.0.
